docs/docs/how-to/plugins-and-themes/adding-gatsby-image-support.md

```diff
@@ -0,0 +1,261 @@
+}
+```
+
+If your host supports it, we recommend using auto-format to deliver next-generation image formats to supported browsers. In this case, ignore the `format` option.
+
+### Create your image data function
+
+This is a similar to the image resolver described above. However because it executes in the browser, the functions that you use should be fast, and synchronous, and you can't use node APIs. You will not be downloading and generating base64 placeholders, for example, or calculating dominant colors. If you have these values pre-calculated then you can pass these in and use them, but they need to be available in the props that you pass to the function at runtime.
+
+The function should accept the props that will be passed into your component, and at a minimum needs to take the props required by the `getImageData` helper function from `gatsby-plugin-image`. Here is an example for an image host:
+
+```js
+import { getImageData } from "gatsby-plugin-image"
+
+export function getExampleImageData({ image, ...props }) {
+  return getImageData({
+    baseUrl: image.url,
+    sourceWidth: image.width,
+    sourceHeight: image.height,
+    urlBuilder,
+    pluginName: "gatsby-source-example",
+    // If your host supports auto-format/content negotiation, pass this as the formats array
+    formats: ["auto"],
+    ...props,
+  })
+}
+```
+
+You can export this function as a public API, and users can use the function to generate data to pass to `GatsbyImage`:
+
+```jsx
+// This might come from an API at runtime
+const image = {
+  url: "kitten.jpg",
+  width: 800,
+  height: 600,
+}
+const imageData = getExampleImageData({ image, layout: "fixed", width: 400 })
+return <GatsbyImage image={imageData} alt="Kitten" />
+```
+
+### Create your wrapper component
+
+This stage is optional: you may prefer to just share the image data function and let your users pass the result to `<GatsbyImage>`, as shown above. However the developer experience is better with a custom image component.
+
+The component should accept the same props as your image data function, and also all of the props for `<GatsbyImage>` too, which it can pass down to that component. Here's how you might type the props in TypeScript:
+
+```typescript
+interface ImageComponentProps
+  //This is the type for your image data function
+  extends GetGatsbyImageDataProps,
+    // We omit "image" because that's the prop that we generate,
+    Omit<GatsbyImageProps, "image"> {
+  // Any other props can go here
+  myCustomProp?: string
+}
+```
+
+Your component can accept just a URL if that's enough to identify the image, or you could pass a full object. This could be an object that you have passed in through your GraphQL API, or it could be data coming from outside of Gatsby, such as via search results or a shopping cart API. Unlike the GraphQL resolvers or the built-in `StaticImage` component, this can be dynamic data that changes at runtime.
```

Comment:
```suggestion
Your component can accept just a URL if that's enough to identify the image, or you can pass a full object. This can be an object that you have passed in through your GraphQL API, or it can be data coming from outside of Gatsby, such as via search results or a shopping cart API. Unlike the GraphQL resolvers or the built-in `StaticImage` component, this can be dynamic data that changes at runtime.
```